docs/load_balancer.md

```diff
@@ -0,0 +1,278 @@
+so for V1 we will have a single visibility setting.  In future, we may convert this to a variably-typed argument 
+(StringOrStringList, or StringOrVisibility) or just add a richer field.
+
+## API sketch
+
+```
+ServiceSpec:
+  - portalIP
+  - createExternalLoadBalancer  // deprecated
+  - publicIPs[]  // deprecated
+  - visibility // "cluster" or "public" or "loadbalancer"
+  - ports[] {
+      - protocol
+      - port
+      - hostPort  // if non-zero, try.  if visibility == public && hostPort == 0, assign one
+   }
+```
+
+```
+ServiceStatus:
+  - publicIPs[] // mirrors publicIPs in Service today, replaces them in future
+```
+
+We have to differentiate between the load-balancer as seen externally (by the k8s end-user) vs. as seen internally
+(e.g. by kube-proxy).
+For example, if we were using AWS subnets to identify ELBs (which we're not), the external load-balancer name
+that the k8s-end-user cares about would be something like `myelbname.amazonaws.com`, the internal ip-ranges would be
+something like `10.244.250.192/30`.  In the case of GCE these are the same, and it looks like we don't need to track
+an internal name at the moment, but we do not want to bake this assumption into the API.  So publicIPs is _external_.
+
+publicIPs in the ServiceSpec will continue to work as it does today. The user can specify a publicIP to use an existing
+load-balancer/VIP (in practice this only works for OpenStack today).  This will be populated once the load-balancer
+is created.  This can be deprecated in future (once the OpenStack plugin does dynamic VIP creation).
+
+Each cloud-provider can keep track of the mapping from service to load-balancer:
+* GCE uses the name
+* AWS can use name and/or tags
+* OpenStack uses the name
+* Rackspace can use the name and/or metadata
+
+Because the load-balancer IPs can be recreated, we are able to expose the load-balancer information through the status.
+The ServiceStatus will be extended to include publicIPs.  This will be the user-facing IP addresses or hostnames.
+
+If we require additional state in future for kube-proxy routing (for example internal load-balancer IPs),
+that will be added to a new Status field.
+
+By putting this information on Status, an external load-balancer provisioner can use the same mechanism for securely
+updating Status as kube-proxy/kubelet.
+
+
+
+## Creation flow
+
+### visibility: loadbalancer
+
+1. User `POST`s `/api/v1/namespaces/foo/services/bar` `Service{ Spec{ visibility: loadbalancer, ports: [{port: 80}] } }`
+1. API server assigns a hostPort (if the load balancer needs one), accepts and persists
+1. LB Controller wakes up, allocates a load-balancer, and `POST`s `/api/v1/namespaces/foo/services/bar` `Service{ Status{ publicIps: [ "1.2.3.4" ] }}`
+1. Service REST sets `Service.Status.publicIps = [ "1.2.3.4" ]`
+1. kube-proxy wakes up and sets iptables to receive on 1.2.3.4
```

Comment:
Except when the load balancer allocates a DNS name, and it needs to listen on a host port? As written, this seems muddled together with the public case, rather than one being nested in the other.